### PR TITLE
Windows 11 22H2: add good UIA window class names for system tray overflow window and Open With dialog to allow mouse and touch interaction

### LINKED ISSUE #14538: Windows 11 22H2: Open With dialog cannot be interacted with using mouse and/or touch
Hi,

First of a two-part issue set on Windows 11 22H2 and mouse/touch:

### Background:
As pointed out in #13506, some parts of Windows 11 interface cannot be interacted with using mouse and/or touch interaction, with the aforementioned issue dealing with taskbar. It turns out with the introductino of Windows 11 2022 Update 9Version 22H2), more changes were made to the interface such that mouse and/or touch interaction no longer works properly. The latest of this is Open With dialog, with an upcoming change to system tray's overflow list that will be introduced to users soon (next issue), both powered by redesigned UI toolkit.

### Steps to reproduce:
Prerequisites: Windows 11 22H2, NVDA 2022.2 or later (preferably 2022.4 and alpha):
1. Enable mouse mouse tracking (NVDA+M) and report mouse (review cursor settings panel).
2. Open NVDA user configuration folder.
3. Try opening a file that Windows doesn't know how to open, say a pickle file.

### Actual behavior:
When moving the mouse to Open With dialog, NVDA reports "DesktopWindowXamlSource" instead of announcing the dialog content, no such announcement on touchscreen interaction.

### Expected behavior:
NVDA announces Open With dialog content via mouse and/or touch navigation.

#### NVDA installed/portable/running from source:
Installed

#### NVDA version:
Alpha-27514

#### Windows version:
Windows 11 22H2 beta (10.0.22623)

#### Name and version of other software in use when reproducing the issue:
Open With dialog

#### Other information about your system:
This is a Windows Insider Preview machine

#### Does the issue still occur after restarting your computer?
Yes

#### Have you tried any other versions of NVDA? If so, please report their behaviors.
Yes - reproducible in 2022.4

#### If NVDA add-ons are disabled, is your problem still occurring?
Yes

#### Does the issue still occur after you run the COM Registration Fixing Tool in NVDA's tools menu?
Not applicable

### Solution:
Treat Open With top-level window as good UIA window (window class name: Open With). A preliminary solution is part of Windows App Essentials add-on.

### LINKED ISSUE #14539: Windows 11 22H2 (Moment 2): redesigned system tray overflow window cannot be interacted with using mouse and/or touch
Hi,

Follow-up to #14538 but won't show up for a while (Insider Preview feature at the moment):

### Background:
As part of Windows 11 22H2 Moment 2 (under testing), redesigned taskbar and system tray will be made available to users, more so if using tablets such as two-in-one devices. At least taskbar can be navigated using mouse and/or touch, but the redesigned system tray overflow window cannot (at the moment).

### Steps to reproduce:
Prerequisites: Windows 11 22H2 beta builds (one must be a Windows Insider to test this):

1. Install Windows 11 22H2 build 22623 or later (must be a Windows Insider subscribed to beta and/or dev channel).
2. Enable mouse tracking (NVDA+M) and follow mouse from review cursor settings panel.
3. Press Windows+B to move focus to system tray/show hidden icons button.
4. Press Space to activate it.
5. Use the mouse to navigate the overflow window.

### Actual behavior:
No announcement from NVDA regarding overflow buttons.

### Expected behavior:
NVDA announces overflow buttons when moving the mouse.

#### NVDA installed/portable/running from source:
Installed

#### NVDA version:
alpha-27514

#### Windows version:
Windows 11 22H2 beta (10.0.22623)

#### Name and version of other software in use when reproducing the issue:
None

#### Other information about your system:
This is an Insider Preview machine

#### Does the issue still occur after restarting your computer?
Yes

#### Have you tried any other versions of NVDA? If so, please report their behaviors.
Yes - reproducible in 2022.4

#### If NVDA add-ons are disabled, is your problem still occurring?
Yes

#### Does the issue still occur after you run the COM Registration Fixing Tool in NVDA's tools menu?
N/A

### Solution:
Treat the top-level window of the redesigned system tray overflow window as good UIA window (window class name: TopLevelWindowForOverflowXamlIsland). Windows App Essentials add-on includes this solution.

### PR DESCRIPTION
**Pull request by josephsl**
### Link to issue number:
Closes #14538 
Closes #14539 

### Summary of the issue:
Mouse and/or touch interaction does not work properly for Windows 11 22H2 UI, including upcoming redesigned system tray overflow window and Open With dialog.

### Description of user facing changes
Mouse and touch interaction will work when opening system tray overflow window and Open With dialog in Windows 11 2022 Update.

### Description of development approach
Edited "isGoodUIAWindow" found in File Explorer and Open With app modules to classify the following window class names as good UIA windows:

* File Explorer: TopLevelWindowForOverflowXamlIsland (upcoming redesigned system tray overflow window)
* Open With: Open With (top-level window for the dialog)

Along with this change, type information were added to these methods.

### Testing strategy:
Manual testing (prereq: mouse and/or touchscreen; for mouse, mouse tracking should be turned on and NVDA sould be told to follow mouse from review cursor settings panel):

Redesigned system tray overflow window (Windows 11 22H2 build 22623 and later): press Windows +B to move focus to "show hidden icons" button, press Space to expand it, then use the mouse to move around the overflow window (located above the icons button).

Open With dialog: click on a file type that Windows cannot open (such as a pickle file) and use the mosue and/or touch to navigate and interact with the dialog.

### Known issues with pull request:
None

### Change log entries:
Bug fixes (under Windows 11 section):

In Windows 11 22H2, it is again possible to use mouse and touch interaction to navigate and interact in areas such as redesigned system tray overflow window and Open With dialog. (#14538, #14539)

### Code Review Checklist:

- [x] Pull Request description:
  - description is up to date
  - change log entries
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] API is compatible with existing add-ons.
- [x] Documentation:
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] Security precautions taken.

### CODE REVIEW

**Comment by mazen428** [comment]
Hello.
All the windows that block touch or mouse access seem to have the name 'DesktopWindowXamlSource', with a role of unknown or pane. I think it would be better to make a window like this UIA or just transparent to mouse instead of manually adding class names for everything. What about apps from the Microsoft Store that use XAML.

**Comment by josephsl** [comment]
Hi, tried this to no avail at the moment – it does not work for touch users, although further testing is needed to determine what’s going on. Thanks.

**Comment by josephsl** [comment]
Hi,

Just telling NVDA that XAML source windows are UIA won't be enough - it is part of aps (specifically, WinUI/XAML Islands), and it is really the top-level windows that are seen as IAccessible, thus these windows must be the ones reclassified as UIA. We can't also rely on the idea that XAML source windows will have the same window class names across apps - class names are different between File Explorer and Windows Terminal. Most importantly, if XAML source window is reclassified as a UIA element, it leads to NVDA becoming unresponsive at startup for several seconds.

Thanks.

File: source/appModules/explorer.py [hunk 1]
```diff
@@ -504,7 +504,7 @@ def event_gainFocus(self, obj, nextHandler):
 
 		nextHandler()
 
-	def isGoodUIAWindow(self, hwnd):
+	def isGoodUIAWindow(self, hwnd: int) -> bool:
```
Comment (seanbudd) [inline source/appModules/explorer.py:507]:
```suggestion
	from winAPI.types import HWNDValT
	def isGoodUIAWindow(self, hwnd: HWNDValT) -> bool:
```

File: source/appModules/openwith.py [hunk 2]
```diff
@@ -40,9 +40,14 @@ def chooseNVDAObjectOverlayClasses(self, obj, clsList):
 			elif automationId == "ImmersiveOpenWithFlyout":
 				clsList.insert(0, ImmersiveOpenWithFlyout)
 
-	def isGoodUIAWindow(self, hwnd):
-		# #11335: Open With dialog isn't read in Windows 10 Version 2004 (May 2020 Update).
-		# Note that treating the below window as a UIA window will make NVDA no longer announce "pane".
-		if winUser.getClassName(hwnd) == "Shell_Flyout":
+	def isGoodUIAWindow(self, hwnd: int) -> bool:
```
Comment (seanbudd) [inline source/appModules/openwith.py:43]: same here in regards to the hwnd type

**Comment by josephsl** [comment]
Hi, ah, wasn’t sure which type to apply for window handle… Thanks.

File: source/appModules/openwith.py [hunk 3]
```diff
@@ -40,9 +40,15 @@ def chooseNVDAObjectOverlayClasses(self, obj, clsList):
 			elif automationId == "ImmersiveOpenWithFlyout":
 				clsList.insert(0, ImmersiveOpenWithFlyout)
 
-	def isGoodUIAWindow(self, hwnd):
-		# #11335: Open With dialog isn't read in Windows 10 Version 2004 (May 2020 Update).
-		# Note that treating the below window as a UIA window will make NVDA no longer announce "pane".
-		if winUser.getClassName(hwnd) == "Shell_Flyout":
+	from winAPI.types import HWNDValT
```
Comment (seanbudd) [inline source/appModules/openwith.py:43]: You should probably pull these imports out to the top of the file

**Review by seanbudd — APPROVED** [review]

**Comment by Rockstar50373** [comment]
Thanks for the update!

**Pull request merged**